File: JavaScript/1-simple.js
'use strict';

const EventEmitter = function() {
  this.events = {}; // hash of array of function
};

EventEmitter.prototype.on = function(name, fn) {
  const event = this.events[name];
  if (event) event.push(fn);
  else this.events[name] = [fn];
};

EventEmitter.prototype.emit = function(name, ...data) {
  const event = this.events[name];
  if (!event) return;
  for (const listener of event) listener(...data);
};

module.exports = EventEmitter;

File: Go/1-simple.go
package emitter

type Listener = func(...interface{})
type EventEmitter map[string][]Listener

func (ee EventEmitter) On(name string, listener Listener) {
	ee[name] = append(ee[name], listener)
}

func (ee EventEmitter) Emit(name string, data ...interface{}) {
	for _, listener := range ee[name] {
		listener(data...)
	}
}
